lock.go

```diff
@@ -49,10 +50,24 @@ func (l *Lock) Lock() error {
 	return l.LockWithData([]byte{})
 }
 
+// TryLock attempts to acquire the lock before a timeout. It works like LockWithData, but it doesn't
+// write any data to the lock node.
+func (l *Lock) TryLock(timeout time.Duration) error {
+	return l.TryLockWithData([]byte{}, timeout)
+}
+
 // LockWithData attempts to acquire the lock, writing data into the lock node.
 // It will wait to return until the lock is acquired or an error occurs. If
 // this instance already has the lock then ErrDeadlock is returned.
 func (l *Lock) LockWithData(data []byte) error {
+	return l.TryLockWithData(data, time.Duration(-1))
+}
+
+// TryLockWithData attempts to acquire the lock, writing data into the lock node.
+// It will wait to return until the lock is acquired, an error occurs, or the timeout. If
+// this instance already has the lock then ErrDeadlock is returned.  A negative
+// timeout is waiting forever
+func (l *Lock) TryLockWithData(data []byte, timeout time.Duration) error {
```

Comment:
in place of any timeout here it may be best to use context.Context since this function has multiple places where we wait or loop forever. both the unconditional for loop on children, and the channel read from watch events after failing a lock.